dali/operators/generic/histogram/histogram.cc

```diff
@@ -0,0 +1,685 @@
+// Copyright (c) 2021, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/operators/generic/histogram/histogram.h"
+
+#include "dali/kernels/common/copy.h"
+#include "dali/kernels/transpose/transpose.h"
+#include "dali/pipeline/operator/op_schema.h"
+
+
+#include "dali/core/tensor_shape_print.h"
+
+#include "opencv2/imgproc.hpp"
+#include "opencv2/core/core_c.h"
+
+#include <iostream>
+
+using namespace dali;
+using namespace dali::hist_detail;
+
+#define id_(x) x
+
+#define HistogramOpName id_(histogram__Histogram)
+#define UniformHistogramOpName id_(histogram__UniformHistogram)
+
+#define str_next_(x) #x
+#define str_(x) str_next_(x)
+
+namespace {
+
+constexpr const char histogramOpString[] = str_(HistogramOpName);
+constexpr const char unifromHistogramOpString[] = str_(UniformHistogramOpName);
+
+static constexpr int HISTOGRAM_MAX_CH = CV_MAX_DIM;
+
+std::vector<float> GetFlattenedRanges(int sample, const workspace_t<CPUBackend> &ws) {
+  std::vector<float> ranges;
+
+  for (int r = 1; r < ws.NumInput(); ++r) {
+    auto &dim_ranges = ws.template Input<CPUBackend>(r);
+    auto range_view = view<const float>(dim_ranges);
+    for (int i = 0; i < range_view[sample].num_elements(); ++i) {
+      ranges.push_back(range_view.tensor_data(sample)[i]);
+    }
+  }
+  return ranges;
+}
+
+template <typename Ty_>
+struct CVMatType {
+  static int get(int) {
+    DALI_ENFORCE(false, "Unreachable - invalid type");
+  }
+};
```

Comment:
How about getting rid of the specializations and going with:
```suggestion
template <typename ChannelType>
struct CVMatType {
  static int get(int nchannels) {
    return CV_MAKETYPE(cv::DataDepth<ChannelType>>::type, nchannels);
  }
};
```
?

Also, the trailing underscore in template arguments is not a coding style we use.